U/danielsf/speed up chipname

danielsf opened the pull request:
The current chipNameFromPupilCoords() method is unnecessarily general.  It allows for the possibility that a single set of pupil coordinates could fall on more than one detector (in the LSST camera, this is only possible with pairs of wavefront sensors, since one wavefront sensor is kept in focus and the other is kept out of focus).  This generality is realized by looping over every pair of pupill coordinates and detector, to see if the detector contains the pupil coordinate, which slows down the code.  This PR adds a method chipNameFromPupilCoordsLSST() that, for each point in pupil coordinates, only considers detectors that are 1.1 detector radii away from the point.  This simplification achieves a factor of 6 to 7 speed up relative to chipNameFromPupilCoords().

danielsf committed:
add obs_lsstSim to dependency

so we can write methods optimized for searching on the LSST
camera

danielsf committed:
add LsstCameraUtils.py with method to find LSST chip names
(the method is called chipNameFromPupilCoordsLSST)

this is about 3 times faster than just using chipNameFromPupilCoords()

danielsf committed:
added unit test for LsstCameraUtils

danielsf committed:
clarify message when object falls on more than one chip

danielsf committed:
allow objects to fall on more than one wavefront sensor

danielsf committed:
remove obsolete line from error message

danielsf committed:
use an array of ints to determine if a point has a chip

this achieved another factor of 2 speed up

danielsf committed:
reduce radius which makes a chip acceptable to search

danielsf committed:
added ability to put sources on multiple detectors in LsstCameraUtils
(if the wavefront sensors are in play)

danielsf committed:
allow multiple chips for any point that could be on a wavefront sensor

danielsf committed:
fix bug

danielsf committed:
fix spelling error

danielsf committed:
do not force multiple chip names into strings

danielsf committed:
convert to np.arrays on output

danielsf committed:
CameraUtils treat allow_multiple_chips sensibly

danielsf committed:
fix bug

danielsf committed:
added unit test to verify how we handle allow_multiple_chips

danielsf committed:
more helpful warning message

danielsf committed:
fix grammar

danielsf committed:
lsst camera utils handle multiple chips

danielsf committed:
fix grammar in warning message

danielsf committed:
fixed spelling errors

danielsf committed:
fix docstrings in CameraUtils.py

danielsf committed:
specify units in CameraUtils docstrings

danielsf committed:
added _chipNameFromRaDecLSST

danielsf committed:
speed up unit test by testing on fewer (xpupil, ypupil) pairs

danielsf committed:
make sure that _chipNameFromRaDecLSST raises exceptions when it should

danielsf committed:
added chipNameFromRaDecLSST

danielsf committed:
added _pixelCoordsFromRaDecLSST

danielsf committed:
test for null chip name with 'is None' rather than 'not'

danielsf committed:
cache detector systems in pixelCoordsFromPupilCoords

danielsf committed:
test pixelCoordsFromRaDecLSST without distortion

danielsf committed:
test that _pixelCoordsFromRaDec raises exceptions when it should

danielsf committed:
test that _pixelCoordsFromRaDec works with forced chip name

danielsf committed:
add pixelCoordsFromRaDecLSST

danielsf committed:
speed up unit test

danielsf committed:
better way to verify a pt is in play

danielsf committed:
add comments to LsstCameraUtils.py

danielsf committed:
pep8 changes to LsstCameraUtils.py

danielsf committed:
pep8 changes to CameraUtils.py

danielsf committed:
pep8 changes to testLsstCameraUtils.py

yoachim commented:
I'm getting unit test fails of:
```
>       self.assertEqual(len(w_list), 1)
E       AssertionError: 0 != 1
tests/testCameraUtils.py:336: AssertionError
```
and
```
>       self.assertIn(chipB, name[1])
E       AssertionError: 'R:4,0 S:0,2,B' not found in 'R:4,0 S:0,2,A'
tests/testLsstCameraUtils.py:82: AssertionError
```

danielsf committed:
amend comments about multple chips (it might be a bug)

danielsf committed:
remove test_multiple_chips unit tests

it is possible that the "a point on the sky can appear on two wavefront
sensors" behavior is a bug in obs_lsstSim.  If it is not, we can always
revert this commit and restore these tests.

The investigation of the bug is ticket DM-8075

danielsf commented:
I have removed the offending unit tests.  It is unclear whether the behavior they were testing is a bug in obs_lsstSim or not (see ticket DM-8075).  If it is not a bug, we can just revert commits 56756c8 and 4a3de29 and restore the tests.

yoachim commented on python/lsst/sims/coordUtils/LsstCameraUtils.py:
```diff
@@ -0,0 +1,427 @@
+    @return outputNameList is a numpy array of the names of the detectors
+    """
+
+    global _lsst_camera
```
I like avoiding globals by making the object an attribute of the function, but since `_lsst_camera` gets used other places too a global makes more sense.

yoachim commented:
Looks good to merge. If you wanted, you could update to numpy style doc strings.
